File: sdk/python/aegis/_http.py
from __future__ import annotations

import json
import urllib.error
import urllib.request
from collections.abc import Iterator
from typing import Any

from .errors import AegisAPIError, AegisConnectionError, AegisStreamError, map_api_error


class HTTPTransport:
    def __init__(self, *, base_url: str, api_key: str | None, timeout: float) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._timeout = timeout

    def request_json(
        self,
        method: str,
        path: str,
        *,
        payload: dict[str, Any] | None = None,
        authenticated: bool = True,
    ) -> dict[str, Any]:
        request = self._build_request(method, path, payload=payload, authenticated=authenticated)
        try:
            with urllib.request.urlopen(request, timeout=self._timeout) as response:
                body = response.read().decode("utf-8")
        except urllib.error.HTTPError as exc:
            raise self._map_http_error(exc) from exc
        except (urllib.error.URLError, TimeoutError, OSError) as exc:
            raise AegisConnectionError(str(exc)) from exc

        try:
            payload = json.loads(body)
        except json.JSONDecodeError as exc:
            raise AegisConnectionError(f"server returned invalid JSON: {exc}") from exc
        if not isinstance(payload, dict):
            raise AegisConnectionError("server returned non-object JSON")
        return payload

    def stream_sse(
        self,
        path: str,
        *,
        payload: dict[str, Any],
        authenticated: bool = True,
    ) -> Iterator[dict[str, Any]]:
        request = self._build_request("POST", path, payload=payload, authenticated=authenticated)
        request.add_header("Accept", "text/event-stream")
        try:
            response = urllib.request.urlopen(request, timeout=self._timeout)
        except urllib.error.HTTPError as exc:
            raise self._map_http_error(exc) from exc
        except (urllib.error.URLError, TimeoutError, OSError) as exc:
            raise AegisConnectionError(str(exc)) from exc

        content_type = response.headers.get("Content-Type", "")
        if "text/event-stream" not in content_type:
            try:
                body = response.read().decode("utf-8")
            finally:
                response.close()
            raise AegisStreamError(f"expected text/event-stream, got {content_type or 'unknown'}: {body}")

        def iterator() -> Iterator[dict[str, Any]]:
            data_lines: list[str] = []
            try:
                while True:
                    raw = response.readline()
                    if not raw:
                        break
                    line = raw.decode("utf-8").rstrip("\r\n")
                    if not line:
                        if data_lines:
                            payload_text = "\n".join(data_lines)
                            data_lines.clear()
                            try:
                                yield json.loads(payload_text)
                            except json.JSONDecodeError as exc:
                                raise AegisStreamError(f"invalid SSE payload: {exc}") from exc
                        continue
                    if line.startswith(":"):
                        continue
                    if line.startswith("data: "):
                        data_lines.append(line[6:])
                if data_lines:
                    payload_text = "\n".join(data_lines)
                    yield json.loads(payload_text)
            finally:
                response.close()

        return iterator()

    def _build_request(
        self,
        method: str,
        path: str,
        *,
        payload: dict[str, Any] | None,
        authenticated: bool,
    ) -> urllib.request.Request:
        data = None
        headers = {"Accept": "application/json"}
        if payload is not None:
            data = json.dumps(payload).encode("utf-8")
            headers["Content-Type"] = "application/json"
        if authenticated and self._api_key:
            headers["Authorization"] = f"Bearer {self._api_key}"
        return urllib.request.Request(self._base_url + path, data=data, headers=headers, method=method)

    def _map_http_error(self, exc: urllib.error.HTTPError) -> AegisAPIError:
        body = exc.read().decode("utf-8", errors="replace")
        code = "http_error"
        message = body.strip() or exc.reason or f"HTTP {exc.code}"
        details = None
        try:
            payload = json.loads(body)
        except json.JSONDecodeError:
            payload = None
        if isinstance(payload, dict) and isinstance(payload.get("error"), dict):
            error = payload["error"]
            code = str(error.get("code", code))
            message = str(error.get("message", message))
            raw_details = error.get("details")
            details = dict(raw_details) if isinstance(raw_details, dict) else None
        return map_api_error(exc.code, code, message, details)

File: sdk/typescript/src/http.ts
import { AegisConnectionError, AegisStreamError, mapApiError } from "./errors.js";

export class HTTPTransport {
  private readonly baseUrl: string;
  private readonly apiKey?: string;
  private readonly timeoutMs: number;

  constructor(options: { baseUrl: string; apiKey?: string; timeoutMs: number }) {
    this.baseUrl = options.baseUrl.replace(/\/$/, "");
    this.apiKey = options.apiKey;
    this.timeoutMs = options.timeoutMs;
  }

  async requestJson(method: string, path: string, options: { payload?: Record<string, unknown>; authenticated?: boolean } = {}): Promise<Record<string, unknown>> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await fetch(this.baseUrl + path, this.buildRequest(method, options.payload, options.authenticated ?? true, controller.signal));
      if (!response.ok) throw await this.mapHttpError(response);
      const text = await response.text();
      try {
        return JSON.parse(text) as Record<string, unknown>;
      } catch (error) {
        throw new AegisConnectionError(`server returned invalid JSON: ${String(error)}`);
      }
    } catch (error) {
      if (error instanceof Error && error.name === "AbortError") {
        throw new AegisConnectionError("request timed out");
      }
      if (error instanceof Error && error.name.startsWith("Aegis")) {
        throw error;
      }
      throw new AegisConnectionError(String(error));
    } finally {
      clearTimeout(timer);
    }
  }

  async *streamSse(path: string, options: { payload: Record<string, unknown>; authenticated?: boolean }): AsyncIterable<Record<string, unknown>> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    let response: Response;
    try {
      response = await fetch(this.baseUrl + path, this.buildRequest("POST", options.payload, options.authenticated ?? true, controller.signal, true));
    } catch (error) {
      clearTimeout(timer);
      if (error instanceof Error && error.name === "AbortError") {
        throw new AegisConnectionError("request timed out");
      }
      throw new AegisConnectionError(String(error));
    }

    if (!response.ok) {
      clearTimeout(timer);
      throw await this.mapHttpError(response);
    }

    const contentType = response.headers.get("content-type") ?? "";
    if (!contentType.includes("text/event-stream")) {
      clearTimeout(timer);
      const body = await response.text();
      throw new AegisStreamError(`expected text/event-stream, got ${contentType || "unknown"}: ${body}`);
    }

    const reader = response.body?.getReader();
    if (!reader) {
      clearTimeout(timer);
      throw new AegisStreamError("streaming response did not expose a readable body");
    }

    const decoder = new TextDecoder();
    let buffer = "";
    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        buffer += decoder.decode(value, { stream: true });
        let separator = buffer.indexOf("\n\n");
        while (separator !== -1) {
          const chunk = buffer.slice(0, separator);
          buffer = buffer.slice(separator + 2);
          const parsed = parseSseChunk(chunk);
          if (parsed) yield parsed;
          separator = buffer.indexOf("\n\n");
        }
      }
      buffer += decoder.decode();
      const finalChunk = parseSseChunk(buffer.trim());
      if (finalChunk) yield finalChunk;
    } finally {
      clearTimeout(timer);
      reader.releaseLock();
    }
  }

  private buildRequest(method: string, payload: Record<string, unknown> | undefined, authenticated: boolean, signal: AbortSignal, stream = false): RequestInit {
    const headers: Record<string, string> = { Accept: stream ? "text/event-stream" : "application/json" };
    if (payload) headers["Content-Type"] = "application/json";
    if (authenticated && this.apiKey) headers.Authorization = `Bearer ${this.apiKey}`;
    return {
      method,
      headers,
      body: payload ? JSON.stringify(payload) : undefined,
      signal,
    };
  }

  private async mapHttpError(response: Response) {
    const body = await response.text();
    let code = "http_error";
    let message = body.trim() || response.statusText || `HTTP ${response.status}`;
    let details: Record<string, unknown> | undefined;
    try {
      const payload = JSON.parse(body) as { error?: { code?: unknown; message?: unknown; details?: unknown } };
      if (payload.error && typeof payload.error === "object") {
        if (typeof payload.error.code === "string") code = payload.error.code;
        if (typeof payload.error.message === "string") message = payload.error.message;
        if (payload.error.details && typeof payload.error.details === "object") details = payload.error.details as Record<string, unknown>;
      }
    } catch {
      // keep fallback body message
    }
    return mapApiError(response.status, code, message, details);
  }
}

function parseSseChunk(chunk: string): Record<string, unknown> | undefined {
  if (!chunk) return undefined;
  const dataLines: string[] = [];
  for (const rawLine of chunk.split(/\r?\n/)) {
    if (!rawLine || rawLine.startsWith(":")) continue;
    if (rawLine.startsWith("data: ")) dataLines.push(rawLine.slice(6));
  }
  if (dataLines.length === 0) return undefined;
  try {
    return JSON.parse(dataLines.join("\n")) as Record<string, unknown>;
  } catch (error) {
    throw new AegisStreamError(`invalid SSE payload: ${String(error)}`);
  }
}
